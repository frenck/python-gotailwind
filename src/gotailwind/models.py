"""Asynchronous Python client for Tailwind garage door openers."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Generic, Self, TypeVar

from mashumaro import field_options
from mashumaro.config import BaseConfig
from mashumaro.mixins.orjson import DataClassORJSONMixin
from mashumaro.types import SerializationStrategy

from gotailwind.const import (
    MIN_REQUIRED_FIRMWARE_VERSION,
    TailwindDoorOperationCommand,
    TailwindDoorState,
    TailwindRequestType,
    TailwindResponseResult,
)
from gotailwind.exceptions import (
    TailwindAuthenticationError,
    TailwindResponseError,
    TailwindUnsupportedFirmwareVersionError,
)
from gotailwind.util import tailwind_device_id_to_mac_address

_RequestData = TypeVar("_RequestData")
_ResponseT = TypeVar("_ResponseT")


class IntegerIsBoolean(SerializationStrategy):
    """Boolean serialization strategy for integers."""

    def serialize(self, value: bool) -> int:  # noqa: FBT001
        """Serialize a boolean to an integer."""
        return int(value)

    def deserialize(self, value: int) -> bool:
        """Deserialize an integer to a boolean."""
        return bool(value)


class BaseModel(DataClassORJSONMixin):
    """Base model for all Tailwind models."""

    # pylint: disable-next=too-few-public-methods
    class Config(BaseConfig):
        """Mashumaro configuration."""

        serialization_strategy = {bool: IntegerIsBoolean()}  # noqa: RUF012
        serialize_by_alias = True
        omit_none = True


@dataclass(kw_only=True)
class TailwindRequestData(BaseModel):
    """Request object for Tailwind devices."""

    request_type: TailwindRequestType = field(
        init=False,
        default=TailwindRequestType.GET,
        metadata=field_options(alias="type"),
    )
    name: str = field(init=False, default="dev_st")


@dataclass(kw_only=True)
class TailwindRequest(Generic[_RequestData, _ResponseT], BaseModel):
    """Base request object for Tailwind devices."""

    data: _RequestData
    version: str = "0.1"
    product: str | None = None

    @property
    def response_type(self) -> type[_ResponseT]:
        """Return the type of the response."""
        raise NotImplementedError


@dataclass(kw_only=True)
class TailwindResponse(BaseModel):
    """Base response object for Tailwind devices."""

    @classmethod
    def __pre_deserialize__(
        cls: type[Self],
        d: dict[Any, Any],
    ) -> dict[Any, Any]:
        """Raise when response was unexpected."""
        result = d.get("result")
        if result == TailwindResponseResult.OK:
            return d

        if result == TailwindResponseResult.AUTH_ERROR:
            msg = "Authentication error. Provided local token is not valid."
            raise TailwindAuthenticationError(msg)

        raise TailwindResponseError(
            d.get("message", d.get("Info", "Unknown error")),
        )


@dataclass(kw_only=True)
class TailwindDoor(BaseModel):
    """Object holding the door status of a Tailwind connected garage door."""

    disabled: bool
    door_id: str
    index: int
    locked_out: bool = field(metadata=field_options(alias="lockup"))
    state: TailwindDoorState = field(metadata=field_options(alias="status"))


@dataclass(kw_only=True)
class TailwindDeviceStatus(TailwindResponse):
    """Object holding the door status of a Tailwind connected garage doors."""

    device_id: str = field(metadata=field_options(alias="dev_id"))
    firmware_version: str = field(metadata=field_options(alias="fw_ver"))
    night_mode_enabled: bool = field(metadata=field_options(alias="night_mode_en"))
    product: str
    protocol_version: str = field(metadata=field_options(alias="proto_ver"))
    led_brightness: int

    number_of_doors: int = field(metadata=field_options(alias="door_num"))
    doors: dict[str, TailwindDoor] = field(
        metadata=field_options(alias="data"), default_factory=dict
    )

    @classmethod
    def __pre_deserialize__(
        cls: type[TailwindDeviceStatus], d: dict[Any, Any]
    ) -> dict[Any, Any]:
        """Modify data before deserialization."""
        if (version := d.get("fw_ver")) and version < MIN_REQUIRED_FIRMWARE_VERSION:
            msg = (
                f"Unsupported firmware version {version}. "
                f"Minimum required version is {MIN_REQUIRED_FIRMWARE_VERSION}. "
                f"Please update your Tailwind device."
            )
            raise TailwindUnsupportedFirmwareVersionError(msg)

        super().__pre_deserialize__(d)
        for door_id, door in d.get("data", {}).items():
            door["door_id"] = door_id
        return d

    @property
    def mac_address(self) -> str:
        """Return the mac address."""
        return tailwind_device_id_to_mac_address(self.device_id)


@dataclass(kw_only=True)
class TailwindDeviceStatusRequestData(TailwindRequestData):
    """Request object for Tailwind devices."""

    name = "dev_st"


@dataclass(kw_only=True)
class TailwindDeviceStatusRequest(
    TailwindRequest[TailwindDeviceStatusRequestData, TailwindDeviceStatus]
):
    """Request object for Tailwind devices."""

    data: TailwindDeviceStatusRequestData = field(
        default_factory=TailwindDeviceStatusRequestData
    )

    @property
    def response_type(self) -> type[TailwindDeviceStatus]:
        """Return the HTTP method for the request."""
        return TailwindDeviceStatus


@dataclass(kw_only=True)
class TailwindIdentifyRequestData(TailwindRequestData):
    """Request object for Tailwind devices."""

    request_type = TailwindRequestType.SET
    name = "identify"


@dataclass(kw_only=True)
class TailwindIdentifyRequest(
    TailwindRequest[TailwindIdentifyRequestData, TailwindResponse]
):
    """Request object for Tailwind devices."""

    data: TailwindIdentifyRequestData = field(
        default_factory=TailwindIdentifyRequestData
    )

    @property
    def response_type(self) -> type[TailwindResponse]:
        """Return the HTTP method for the request."""
        return TailwindResponse


@dataclass(kw_only=True)
class TailwindDoorOperationRequestDataValue(BaseModel):
    """Request object for Tailwind devices."""

    index: int = field(metadata=field_options(alias="door_idx"))
    operation: TailwindDoorOperationCommand = field(metadata=field_options(alias="cmd"))


@dataclass(kw_only=True)
class TailwindDoorOperationRequestData(TailwindRequestData):
    """Request object for Tailwind devices."""

    value: TailwindDoorOperationRequestDataValue
    request_type = TailwindRequestType.SET
    name = "door_op"


@dataclass(kw_only=True)
class TailwindDoorOperationRequest(
    TailwindRequest[TailwindDoorOperationRequestData, TailwindResponse]
):
    """Request object for Tailwind devices."""

    data: TailwindDoorOperationRequestData

    @property
    def response_type(self) -> type[TailwindResponse]:
        """Return the HTTP method for the request."""
        return TailwindResponse


@dataclass(kw_only=True)
class TailwindLEDBrightnessRequestDataValue(BaseModel):
    """Request object for Tailwind devices."""

    brightness: int


@dataclass(kw_only=True)
class TailwindLEDBrightnessRequestData(TailwindRequestData):
    """Request object for Tailwind devices."""

    value: TailwindLEDBrightnessRequestDataValue
    request_type = TailwindRequestType.SET
    name = "status_led"


@dataclass(kw_only=True)
class TailwindLEDBrightnessRequest(
    TailwindRequest[TailwindLEDBrightnessRequestData, TailwindResponse]
):
    """Request object for Tailwind devices."""

    data: TailwindLEDBrightnessRequestData

    @property
    def response_type(self) -> type[TailwindResponse]:
        """Return the HTTP method for the request."""
        return TailwindResponse
