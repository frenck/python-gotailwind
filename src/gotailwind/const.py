"""Asynchronous Python client for Tailwind garage door openers."""

from enum import StrEnum, auto

from awesomeversion import AwesomeVersion, AwesomeVersionStrategy

OPERATION_WAIT_CYCLES = 120
OPERATION_CYCLE_WAIT = 0.5

MIN_REQUIRED_FIRMWARE_VERSION = AwesomeVersion(
    "10.10", ensure_strategy=AwesomeVersionStrategy.SIMPLEVER
)


class TailwindDoorState(StrEnum):
    """Enum for door state."""

    CLOSED = "close"
    OPEN = "open"


class TailwindDoorOperationCommand(StrEnum):
    """Enum for door operation."""

    CLOSE = auto()
    OPEN = auto()


class TailwindResponseResult(StrEnum):
    """Enum for different response types."""

    OK = "OK"
    ERROR = "Fail"
    AUTH_ERROR = "token fail"


class TailwindRequestType(StrEnum):
    """Enum for different request types."""

    SET = "set"
    GET = "get"
