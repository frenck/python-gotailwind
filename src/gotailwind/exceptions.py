"""Asynchronous Python client for Tailwind garage door openers."""


class TailwindError(Exception):
    """Generic Tailwind exception."""


class TailwindConnectionError(TailwindError):
    """Tailwind connection exception."""


class TailwindConnectionTimeoutError(TailwindConnectionError):
    """Tailwind connection timeout exception."""


class TailwindResponseError(TailwindError):
    """Tailwind unexpected response exception."""


class TailwindAuthenticationError(TailwindResponseError):
    """Tailwind connection exception."""


class TailwindDoorUnknownError(TailwindError):
    """Tailwind unknown door exception."""


class TailwindDoorOperationError(TailwindError):
    """Tailwind door operation exception."""


class TailwindDoorDisabledError(TailwindDoorOperationError):
    """Tailwind disabled door exception."""


class TailwindDoorLockedOutError(TailwindDoorOperationError):
    """Tailwind locked out door exception."""


class TailwindUnsupportedFirmwareVersionError(TailwindError):
    """Tailwind unsupported version exception."""
