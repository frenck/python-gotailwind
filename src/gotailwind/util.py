"""Asynchronous Python client for Tailwind garage door openers."""


def tailwind_device_id_to_mac_address(device_id: str) -> str:
    """Convert a Tailwind device ID to a MAC address."""
    return ":".join([part.zfill(2) for part in device_id.strip("_").split("_")])
