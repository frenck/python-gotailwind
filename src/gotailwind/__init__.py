"""Asynchronous Python client for Tailwind garage door openers."""
from .const import (
    MIN_REQUIRED_FIRMWARE_VERSION,
    TailwindDoorOperationCommand,
    TailwindDoorState,
    TailwindRequestType,
    TailwindResponseResult,
)
from .exceptions import (
    TailwindAuthenticationError,
    TailwindConnectionError,
    TailwindConnectionTimeoutError,
    TailwindDoorDisabledError,
    TailwindDoorLockedOutError,
    TailwindDoorOperationError,
    TailwindDoorUnknownError,
    TailwindError,
    TailwindResponseError,
    TailwindUnsupportedFirmwareVersionError,
)
from .models import (
    TailwindDeviceStatus,
    TailwindDeviceStatusRequest,
    TailwindDeviceStatusRequestData,
    TailwindDoor,
    TailwindDoorOperationRequest,
    TailwindDoorOperationRequestData,
    TailwindDoorOperationRequestDataValue,
    TailwindIdentifyRequest,
    TailwindIdentifyRequestData,
    TailwindLEDBrightnessRequest,
    TailwindLEDBrightnessRequestData,
    TailwindLEDBrightnessRequestDataValue,
    TailwindRequest,
    TailwindRequestData,
    TailwindResponse,
)
from .tailwind import Tailwind
from .util import tailwind_device_id_to_mac_address

__all__ = [
    "MIN_REQUIRED_FIRMWARE_VERSION",
    "tailwind_device_id_to_mac_address",
    "Tailwind",
    "TailwindAuthenticationError",
    "TailwindConnectionError",
    "TailwindConnectionTimeoutError",
    "TailwindDeviceStatus",
    "TailwindDeviceStatusRequest",
    "TailwindDeviceStatusRequestData",
    "TailwindDoor",
    "TailwindDoorDisabledError",
    "TailwindDoorLockedOutError",
    "TailwindDoorOperationCommand",
    "TailwindDoorOperationError",
    "TailwindDoorOperationRequest",
    "TailwindDoorOperationRequestData",
    "TailwindDoorOperationRequestDataValue",
    "TailwindDoorState",
    "TailwindDoorUnknownError",
    "TailwindError",
    "TailwindIdentifyRequest",
    "TailwindIdentifyRequestData",
    "TailwindLEDBrightnessRequest",
    "TailwindLEDBrightnessRequestData",
    "TailwindLEDBrightnessRequestDataValue",
    "TailwindRequest",
    "TailwindRequestData",
    "TailwindRequestType",
    "TailwindResponse",
    "TailwindResponseError",
    "TailwindResponseResult",
    "TailwindUnsupportedFirmwareVersionError",
]
