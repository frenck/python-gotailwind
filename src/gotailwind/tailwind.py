"""Asynchronous Python client for Tailwind garage door openers."""
from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from typing import Any, Self, TypeVar

import backoff
import orjson
from aiohttp.client import ClientError, ClientSession
from aiohttp.hdrs import METH_POST
from yarl import URL

from gotailwind.const import (
    OPERATION_CYCLE_WAIT,
    OPERATION_WAIT_CYCLES,
    TailwindDoorOperationCommand,
    TailwindDoorState,
)

from .exceptions import (
    TailwindConnectionError,
    TailwindConnectionTimeoutError,
    TailwindDoorDisabledError,
    TailwindDoorLockedOutError,
    TailwindDoorOperationError,
    TailwindDoorUnknownError,
)
from .models import (
    TailwindDeviceStatus,
    TailwindDeviceStatusRequest,
    TailwindDoor,
    TailwindDoorOperationRequest,
    TailwindDoorOperationRequestData,
    TailwindDoorOperationRequestDataValue,
    TailwindIdentifyRequest,
    TailwindLEDBrightnessRequest,
    TailwindLEDBrightnessRequestData,
    TailwindLEDBrightnessRequestDataValue,
    TailwindRequest,
)

_TailwindRequestT = TypeVar("_TailwindRequestT", bound=TailwindRequest[Any, Any])


@dataclass(kw_only=True)
class Tailwind:
    """Main class for handling connections with a Tailwind garage door opener."""

    host: str
    token: str
    request_timeout: float = 8
    session: ClientSession | None = None

    product: str = "iQ3"
    version: str = "0.1"

    _close_session: bool = False

    def __post_init__(self) -> None:
        """Initialize the Tailwind object."""
        self.url = URL.build(scheme="http", host=self.host, path="/json")

    @backoff.on_exception(
        backoff.expo,
        TailwindConnectionError,
        max_tries=3,
        logger=None,
    )
    async def request(
        self,
        request: _TailwindRequestT,
    ) -> _TailwindRequestT.response_type:  # type: ignore[name-defined]
        """Handle a request to a Tailwind device."""
        if self.session is None:
            self.session = ClientSession(json_serialize=orjson.dumps)  # type: ignore[arg-type]
            self._close_session = True
        try:
            async with asyncio.timeout(self.request_timeout):
                response = await self.session.request(
                    METH_POST,
                    self.url,
                    data=request.to_json(),
                    headers={"TOKEN": self.token},
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to the Tailwind device"
            raise TailwindConnectionTimeoutError(msg) from exception
        except (
            ClientError,
            socket.gaierror,
        ) as exception:
            msg = "Error occurred while communicating to the Tailwind device"
            raise TailwindConnectionError(msg) from exception

        response_text = await response.text()
        return request.response_type.from_json(response_text)

    async def status(self) -> TailwindDeviceStatus:
        """Get the current status of the Tailwind device."""
        return await self.request(TailwindDeviceStatusRequest())  # type: ignore[no-any-return]

    async def door_status(self, *, door: int | str) -> TailwindDoor:
        """Get the door object for the given door id."""
        status = await self.status()
        if not (
            door_status := next(
                (
                    door_status
                    for door_status in status.doors.values()
                    if door in (door_status.index, door_status.door_id)
                ),
                None,
            )
        ):
            msg = f"Door {door} not found"
            raise TailwindDoorUnknownError(msg)
        return door_status

    async def identify(self) -> None:
        """Identify the Tailwind device."""
        await self.request(TailwindIdentifyRequest())

    async def status_led(self, *, brightness: int) -> None:
        """Configure the status LED of the Tailwind device."""
        await self.request(
            TailwindLEDBrightnessRequest(
                data=TailwindLEDBrightnessRequestData(
                    value=TailwindLEDBrightnessRequestDataValue(
                        brightness=brightness,
                    )
                )
            )
        )

    async def operate(
        self,
        *,
        door: int | str | TailwindDoor,
        operation: TailwindDoorOperationCommand,
    ) -> TailwindDoor:
        """Open the garage door."""
        if isinstance(door, TailwindDoor):
            door = door.index
        door_status = await self.door_status(door=door)

        if door_status.locked_out:
            msg = f"Door {door} is locked out"
            raise TailwindDoorLockedOutError(msg)

        if door_status.disabled:
            msg = f"Door {door} is disabled"
            raise TailwindDoorDisabledError(msg)

        if (
            operation == TailwindDoorOperationCommand.OPEN
            and door_status.state == TailwindDoorState.OPEN
        ) or (
            operation == TailwindDoorOperationCommand.CLOSE
            and door_status.state == TailwindDoorState.CLOSED
        ):
            msg = f"Door {door} is already in the requested state"
            raise TailwindDoorOperationError(msg)

        await self.request(
            TailwindDoorOperationRequest(
                data=TailwindDoorOperationRequestData(
                    value=TailwindDoorOperationRequestDataValue(
                        index=door_status.index,
                        operation=operation,
                    )
                ),
            ),
        )

        # Wait for the door to open/close
        for _ in range(OPERATION_WAIT_CYCLES):
            await asyncio.sleep(OPERATION_CYCLE_WAIT)
            if (door_status := await self.door_status(door=door)) and (
                (
                    operation == TailwindDoorOperationCommand.CLOSE
                    and door_status.state == TailwindDoorState.CLOSED
                )
                or (
                    operation == TailwindDoorOperationCommand.OPEN
                    and door_status.state == TailwindDoorState.OPEN
                )
            ):
                return door_status

        return door_status

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The Tailwind object.
        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
        await self.close()
