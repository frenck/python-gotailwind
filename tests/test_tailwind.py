"""Asynchronous Python client for Tailwind garage door openers."""

from __future__ import annotations

import asyncio
from unittest.mock import patch

import pytest
from aiohttp import ClientResponse, ClientSession
from aresponses import Response, ResponsesMockServer
from syrupy.assertion import SnapshotAssertion

from gotailwind.const import TailwindDoorOperationCommand, TailwindDoorState
from gotailwind.exceptions import (
    TailwindAuthenticationError,
    TailwindConnectionError,
    TailwindDoorOperationError,
    TailwindDoorUnknownError,
    TailwindResponseError,
    TailwindUnsupportedFirmwareVersionError,
)
from gotailwind.models import TailwindIdentifyRequest
from gotailwind.tailwind import Tailwind

from . import load_fixture


async def test_status(
    aresponses: ResponsesMockServer, snapshot: SnapshotAssertion
) -> None:
    """Test getting status from Tailwind device."""
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("device_status_open.json"),
        ),
    )
    async with Tailwind(host="example.com", token="12346") as tailwind:
        status = await tailwind.status()

    assert status == snapshot(name="object")
    assert status.to_json() == snapshot(name="json")
    assert status.mac_address == "3c:e9:0e:6d:21:84"


async def test_unsupported_firmware_version(aresponses: ResponsesMockServer) -> None:
    """Test trying to use an unsupported firmware version."""
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("device_status_unsupported.json"),
        ),
    )
    async with Tailwind(host="example.com", token="12346") as tailwind:
        with pytest.raises(
            TailwindUnsupportedFirmwareVersionError,
            match="Unsupported firmware version",
        ):
            await tailwind.status()


@pytest.mark.parametrize("door", [1, "door1"])
async def test_door_status(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    door: int | str,
) -> None:
    """Test getting door status from Tailwind device."""

    async def response_handler(request: ClientResponse) -> Response:
        """Response handler for this test."""
        data = await request.json()
        assert data == snapshot(name="request")
        return aresponses.Response(
            status=200,
            text=load_fixture("device_status_open.json"),
        )

    aresponses.add(
        "example.com",
        "/json",
        "POST",
        response_handler,
    )
    async with Tailwind(host="example.com", token="12346") as tailwind:
        status = await tailwind.door_status(door=door)

    assert status == snapshot(name="response")


async def test_door_status_unknown_door(
    aresponses: ResponsesMockServer,
) -> None:
    """Test getting non-existing door status from Tailwind device."""
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("device_status_open.json"),
        ),
    )
    async with Tailwind(host="example.com", token="12346") as tailwind:
        with pytest.raises(TailwindDoorUnknownError, match="Door 42 not found"):
            await tailwind.door_status(door="42")


async def test_identify(
    aresponses: ResponsesMockServer, snapshot: SnapshotAssertion
) -> None:
    """Test the identify method."""

    async def response_handler(request: ClientResponse) -> Response:
        """Response handler for this test."""
        data = await request.json()
        assert data == snapshot(name="request")
        return aresponses.Response(
            status=200,
            text=load_fixture("ok_response.json"),
        )

    aresponses.add(
        "example.com",
        "/json",
        "POST",
        response_handler,
    )
    async with Tailwind(host="example.com", token="12346") as tailwind:
        await tailwind.identify()


async def test_status_led(
    aresponses: ResponsesMockServer, snapshot: SnapshotAssertion
) -> None:
    """Test the status led method."""

    async def response_handler(request: ClientResponse) -> Response:
        """Response handler for this test."""
        data = await request.json()
        assert data == snapshot(name="request")
        return aresponses.Response(
            status=200,
            text=load_fixture("ok_response.json"),
        )

    aresponses.add(
        "example.com",
        "/json",
        "POST",
        response_handler,
    )
    async with Tailwind(host="example.com", token="12346") as tailwind:
        await tailwind.status_led(brightness=42)


async def test_operate_open(
    aresponses: ResponsesMockServer, snapshot: SnapshotAssertion
) -> None:
    """Test operating the Tailwind doors."""

    async def response_handler(request: ClientResponse) -> Response:
        """Response handler for this test."""
        data = await request.json()
        assert data == snapshot(name="request")
        return aresponses.Response(
            status=200,
            text=load_fixture("ok_response.json"),
        )

    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("device_status_closed.json"),
        ),
    )
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        response_handler,
    )
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("device_status_closed.json"),
        ),
        repeat=2,
    )
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("device_status_open.json"),
        ),
    )
    async with Tailwind(host="example.com", token="12346") as tailwind:
        status = await tailwind.operate(
            door="door1", operation=TailwindDoorOperationCommand.OPEN
        )

    assert status == snapshot(name="response")


async def test_operate_close(
    aresponses: ResponsesMockServer, snapshot: SnapshotAssertion
) -> None:
    """Test operating the Tailwind doors."""

    async def response_handler(request: ClientResponse) -> Response:
        """Response handler for this test."""
        data = await request.json()
        assert data == snapshot(name="request")
        return aresponses.Response(
            status=200,
            text=load_fixture("ok_response.json"),
        )

    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("device_status_open.json"),
        ),
    )
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        response_handler,
    )
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("device_status_open.json"),
        ),
        repeat=2,
    )
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("device_status_closed.json"),
        ),
    )
    async with Tailwind(host="example.com", token="12346") as tailwind:
        status = await tailwind.operate(
            door="door1", operation=TailwindDoorOperationCommand.CLOSE
        )
        assert status == snapshot(name="response")


async def test_operate_took_to_long(
    aresponses: ResponsesMockServer, snapshot: SnapshotAssertion
) -> None:
    """Test operating the Tailwind doors."""

    async def response_handler(request: ClientResponse) -> Response:
        """Response handler for this test."""
        data = await request.json()
        assert data == snapshot(name="request")
        return aresponses.Response(
            status=200,
            text=load_fixture("ok_response.json"),
        )

    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("device_status_open.json"),
        ),
    )
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        response_handler,
    )
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("device_status_open.json"),
        ),
        repeat=2,
    )
    async with Tailwind(host="example.com", token="12346") as tailwind:
        with patch("gotailwind.tailwind.OPERATION_WAIT_CYCLES", 1):
            status = await tailwind.operate(
                door="door1", operation=TailwindDoorOperationCommand.CLOSE
            )

    # Reported state should still be open.
    assert status.state == TailwindDoorState.OPEN


async def test_operate_disabled(
    aresponses: ResponsesMockServer,
) -> None:
    """Test operating the Tailwind doors that are disabled."""
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("device_status_disabled.json"),
        ),
    )
    async with Tailwind(host="example.com", token="12346") as tailwind:
        with pytest.raises(TailwindDoorOperationError, match="is disabled"):
            await tailwind.operate(
                door="door1", operation=TailwindDoorOperationCommand.OPEN
            )


async def test_operate_locked_out(
    aresponses: ResponsesMockServer,
) -> None:
    """Test operating the Tailwind doors that are in a locked out state."""
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("device_status_locked_out.json"),
        ),
    )
    async with Tailwind(host="example.com", token="12346") as tailwind:
        with pytest.raises(TailwindDoorOperationError, match="is locked out"):
            await tailwind.operate(
                door="door1", operation=TailwindDoorOperationCommand.OPEN
            )


@pytest.mark.parametrize(
    ("fixture", "operation"),
    [
        ("device_status_open.json", TailwindDoorOperationCommand.OPEN),
        ("device_status_closed.json", TailwindDoorOperationCommand.CLOSE),
    ],
)
async def test_operate_already_in_state(
    aresponses: ResponsesMockServer,
    fixture: str,
    operation: TailwindDoorOperationCommand,
) -> None:
    """Test operating the Tailwind doors that are already in the requested state."""
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture(fixture),
        ),
    )
    async with Tailwind(host="example.com", token="12346") as tailwind:
        with pytest.raises(
            TailwindDoorOperationError, match="already in the requested state"
        ):
            await tailwind.operate(door="door1", operation=operation)


async def test_operate_with_door_object(
    aresponses: ResponsesMockServer,
) -> None:
    """Test opering with a door status object works."""
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("device_status_locked_out.json"),
        ),
        repeat=2,
    )
    async with Tailwind(host="example.com", token="12346") as tailwind:
        door_status = await tailwind.door_status(door="door1")
        with pytest.raises(TailwindDoorOperationError, match="is locked out"):
            await tailwind.operate(
                door=door_status, operation=TailwindDoorOperationCommand.OPEN
            )


async def test_request_with_shared_session(aresponses: ResponsesMockServer) -> None:
    """Test a passed in shared session works as expected."""
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("ok_response.json"),
        ),
    )
    async with ClientSession() as session:
        tailwind = Tailwind(host="example.com", token="123456", session=session)
        await tailwind.request(TailwindIdentifyRequest())
        await tailwind.close()


async def test_timeout(aresponses: ResponsesMockServer) -> None:
    """Test request timeout."""

    # Faking a timeout by sleeping
    async def response_handler(_: ClientResponse) -> Response:
        """Response handler for this test."""
        await asyncio.sleep(8)
        return aresponses.Response(
            status=200,
            text=load_fixture("ok_response.json"),
        )

    aresponses.add(
        "example.com",
        "/json",
        "POST",
        response_handler,
    )
    async with Tailwind(
        host="example.com",
        token="123456",
        request_timeout=1,
    ) as tailwind:
        with pytest.raises(TailwindConnectionError):
            await tailwind.request(TailwindIdentifyRequest())


async def test_http_error400(aresponses: ResponsesMockServer) -> None:
    """Test HTTP 404 response handling."""
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(text="OMG PUPPIES!", status=404),
    )

    async with Tailwind(host="example.com", token="123456") as tailwind:
        with pytest.raises(TailwindConnectionError):
            await tailwind.request(TailwindIdentifyRequest())


async def test_unauthenticated_response(aresponses: ResponsesMockServer) -> None:
    """Test authentication failure."""
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("token_fail.json"),
        ),
    )

    async with Tailwind(host="example.com", token="123456") as tailwind:
        with pytest.raises(TailwindAuthenticationError):
            await tailwind.request(TailwindIdentifyRequest())


async def test_error_response(aresponses: ResponsesMockServer) -> None:
    """Test error failure."""
    aresponses.add(
        "example.com",
        "/json",
        "POST",
        aresponses.Response(
            status=200,
            text=load_fixture("fail_response.json"),
        ),
    )

    async with Tailwind(host="example.com", token="123456") as tailwind:
        with pytest.raises(TailwindResponseError, match="OMG Puppies!"):
            await tailwind.request(TailwindIdentifyRequest())
