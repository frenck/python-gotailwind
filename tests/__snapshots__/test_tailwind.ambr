# serializer version: 1
# name: test_door_status[1][request]
  dict({
    'data': dict({
      'name': 'dev_st',
      'type': 'get',
    }),
    'version': '0.1',
  })
# ---
# name: test_door_status[1][response]
  TailwindDoor(disabled=False, door_id='door2', index=1, locked_out=False, state=<TailwindDoorState.OPEN: 'open'>)
# ---
# name: test_door_status[door1][request]
  dict({
    'data': dict({
      'name': 'dev_st',
      'type': 'get',
    }),
    'version': '0.1',
  })
# ---
# name: test_door_status[door1][response]
  TailwindDoor(disabled=False, door_id='door1', index=0, locked_out=False, state=<TailwindDoorState.OPEN: 'open'>)
# ---
# name: test_identify[request]
  dict({
    'data': dict({
      'name': 'identify',
      'type': 'set',
    }),
    'version': '0.1',
  })
# ---
# name: test_operate_close[request]
  dict({
    'data': dict({
      'name': 'door_op',
      'type': 'set',
      'value': dict({
        'cmd': 'close',
        'door_idx': 0,
      }),
    }),
    'version': '0.1',
  })
# ---
# name: test_operate_close[response]
  TailwindDoor(disabled=False, door_id='door1', index=0, locked_out=False, state=<TailwindDoorState.CLOSED: 'close'>)
# ---
# name: test_operate_open[request]
  dict({
    'data': dict({
      'name': 'door_op',
      'type': 'set',
      'value': dict({
        'cmd': 'open',
        'door_idx': 0,
      }),
    }),
    'version': '0.1',
  })
# ---
# name: test_operate_open[response]
  TailwindDoor(disabled=False, door_id='door1', index=0, locked_out=False, state=<TailwindDoorState.OPEN: 'open'>)
# ---
# name: test_operate_took_to_long[request]
  dict({
    'data': dict({
      'name': 'door_op',
      'type': 'set',
      'value': dict({
        'cmd': 'close',
        'door_idx': 0,
      }),
    }),
    'version': '0.1',
  })
# ---
# name: test_status[json]
  '{"dev_id":"_3c_e9_e_6d_21_84_","fw_ver":"10.10","night_mode_en":0,"product":"iQ3","proto_ver":"0.1","led_brightness":100,"door_num":2,"data":{"door1":{"disabled":0,"door_id":"door1","index":0,"lockup":0,"status":"open"},"door2":{"disabled":0,"door_id":"door2","index":1,"lockup":0,"status":"open"}}}'
# ---
# name: test_status[object]
  TailwindDeviceStatus(device_id='_3c_e9_e_6d_21_84_', firmware_version='10.10', night_mode_enabled=False, product='iQ3', protocol_version='0.1', led_brightness=100, number_of_doors=2, doors={'door1': TailwindDoor(disabled=False, door_id='door1', index=0, locked_out=False, state=<TailwindDoorState.OPEN: 'open'>), 'door2': TailwindDoor(disabled=False, door_id='door2', index=1, locked_out=False, state=<TailwindDoorState.OPEN: 'open'>)})
# ---
# name: test_status_led[request]
  dict({
    'data': dict({
      'name': 'status_led',
      'type': 'set',
      'value': dict({
        'brightness': 42,
      }),
    }),
    'version': '0.1',
  })
# ---
